Mathlib/FieldTheory/KummerExtension.lean

```diff
@@ -0,0 +1,444 @@
+We first develop the theory for a specific `K[n√a] := AdjoinRoot (X ^ n - C a)`.
+The main result is the description of the galois group: `autAdjoinRootXPowSubCEquiv`.
+-/
+
+variable (n : ℕ) (hζ : (primitiveRoots n K).Nonempty) (a : K) (H : Irreducible (X ^ n - C a))
+
+set_option quotPrecheck false in
+scoped[KummerExtension] notation3 "K[" n "√" a "]" => AdjoinRoot (Polynomial.X ^ n - Polynomial.C a)
+
+attribute [nolint docBlame] KummerExtension.«termK[_√_]»
+
+open scoped KummerExtension
+
+section AdjoinRoot
+
+/-- Also see `Polynomial.separable_X_pow_sub_C_unit` -/
+theorem Polynomial.separable_X_pow_sub_C_of_irreducible : (X ^ n - C a).Separable := by
+  letI := Fact.mk H
+  letI : Algebra K K[n√a] := inferInstance
+  have hn := Nat.pos_iff_ne_zero.mpr (ne_zero_of_irreducible_X_pow_sub_C H)
+  by_cases hn' : n = 1
+  · rw [hn', pow_one]; exact separable_X_sub_C
+  have ⟨ζ, hζ⟩ := hζ
+  rw [mem_primitiveRoots (Nat.pos_of_ne_zero <| ne_zero_of_irreducible_X_pow_sub_C H)] at hζ
+  rw [← separable_map (algebraMap K K[n√a]), Polynomial.map_sub, Polynomial.map_pow, map_C, map_X,
+    algebraMap_eq, X_pow_sub_C_eq_prod (hζ.map_of_injective (algebraMap K _).injective) hn
+    (root_X_pow_sub_C_pow n a), separable_prod_X_sub_C_iff']
+  refine injOn_pow_mul_of_isPrimitiveRoot hn (hζ.map_of_injective (algebraMap K _).injective)
+    (root_X_pow_sub_C_ne_zero ?_ _)
+  exact lt_of_le_of_ne (show 1 ≤ n from hn) (Ne.symm hn')
+
+/-- The natural embedding of the roots of unity of `K` into `Gal(K[ⁿ√a]/K)`, by sending
+`η ↦ (ⁿ√a ↦ η • ⁿ√a)`. Also see `autAdjoinRootXPowSubC` for the `AlgEquiv` version. -/
+noncomputable
+def autAdjoinRootXPowSubCHom (hn : 0 < n) :
+    rootsOfUnity ⟨n, hn⟩ K →* (K[n√a] →ₐ[K] K[n√a]) where
+  toFun := fun η ↦ liftHom (X ^ n - C a) (((η : Kˣ) : K) • (root _) : K[n√a]) <| by
+    have := (mem_rootsOfUnity' _ _).mp η.prop
+    dsimp at this
+    rw [map_sub, map_pow, aeval_C, aeval_X, Algebra.smul_def, mul_pow, root_X_pow_sub_C_pow,
+      AdjoinRoot.algebraMap_eq, ← map_pow, this, map_one, one_mul, sub_self]
+  map_one' := algHom_ext <| by simp
+  map_mul' := fun ε η ↦ algHom_ext <| by simp [mul_smul, smul_comm ((ε : Kˣ) : K)]
+
+/-- The natural embedding of the roots of unity of `K` into `Gal(K[ⁿ√a]/K)`, by sending
+`η ↦ (ⁿ√a ↦ η • ⁿ√a)`. -/
+noncomputable
+def autAdjoinRootXPowSubC (hn : 0 < n) :
+    rootsOfUnity ⟨n, hn⟩ K →* (K[n√a] ≃ₐ[K] K[n√a]) :=
+  (AlgEquiv.algHomUnitsEquiv _ _).toMonoidHom.comp (autAdjoinRootXPowSubCHom n a hn).toHomUnits
+
+lemma autAdjoinRootXPowSubC_root (hn : 0 < n) (η) :
+    autAdjoinRootXPowSubC n a hn η (root _) = ((η : Kˣ) : K) • root _ := by
+  dsimp [autAdjoinRootXPowSubC, autAdjoinRootXPowSubCHom, AlgEquiv.algHomUnitsEquiv]
+  apply liftHom_root
+
+variable {n} {a}
+
+/-- The inverse function of `autAdjoinRootXPowSubC` if `K` has all roots of unity.
+See `autAdjoinRootXPowSubCEquiv`. -/
+noncomputable
+def AdjoinRootXPowSubCEquivToRootsOfUnity (hn : 0 < n) (σ : K[n√a] ≃ₐ[K] K[n√a]) :
```

Comment:
Can you write the definition in term mode, leaving only the proof part in tactic mode?